verify_sna/src/main/java/com/twilio/verify_sna/domain/requestmanager/RequestManager.kt

```diff
@@ -74,31 +78,80 @@ class ConcreteRequestManager(
     }
   }
 
+  private fun performRequest(
+    url: String,
+    network: Network,
+    continuation: Continuation<NetworkRequestResult>,
+    connectivityManager: ConnectivityManager,
+    networkCallback: NetworkCallback
+  ) {
+    try {
+      val networkRequestResult = networkRequestProvider.performRequest(url, network)
+      continuation.resume(networkRequestResult)
+    } catch (networkRequestException: TwilioVerifySnaException.NetworkRequestException) {
+      continuation.resumeWithException(
+        networkRequestException
+      )
+    } catch (e: Exception) {
+      continuation.resumeWithException(
+        TwilioVerifySnaException.NetworkRequestException(e)
+      )
+    } finally {
+      connectivityManager.unregisterNetworkCallback(networkCallback)
+    }
+  }
+
   private fun establishCellularConnection(
     connectivityManager: ConnectivityManager,
     url: String,
     continuation: Continuation<NetworkRequestResult>
   ) {
-    val networkRequest = NetworkRequest.Builder()
+    val networkRequestBuilder = NetworkRequest.Builder()
       .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
       .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
-      .build()
-    connectivityManager.requestNetwork(
-      networkRequest,
-      object : NetworkCallback() {
-        override fun onAvailable(network: Network) {
-          try {
-            val networkRequestResult = networkRequestProvider.performRequest(url, network)
-            continuation.resume(networkRequestResult)
-          } catch (networkRequestException: TwilioVerifySnaException.NetworkRequestException) {
+      .addCapability(NetworkCapabilities.NET_CAPABILITY_NOT_RESTRICTED)
+
+    val networkRequest = networkRequestBuilder.build()
+    val networkCallback = object : NetworkCallback() {
+      override fun onAvailable(network: Network) {
+        super.onAvailable(network)
+        if (VERSION.SDK_INT < VERSION_CODES.M) {
+          performRequest(url, network, continuation, connectivityManager, this)
+        }
+      }
+
+      override fun onCapabilitiesChanged(
+        network: Network,
+        networkCapabilities: NetworkCapabilities
+      ) {
+        super.onCapabilitiesChanged(network, networkCapabilities)
+        if (VERSION.SDK_INT >= VERSION_CODES.M) {
+          if (networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)) {
+            performRequest(url, network, continuation, connectivityManager, this)
+          } else {
+            connectivityManager.unregisterNetworkCallback(this)
```

Comment:
capabilities can change multiple times, so this can even be triggered when `NET_CAPABILITY_VALIDATED` is not yet available but there was other change in the network. Can you move the `ungeristerNetworkCallback` and `resumeWIthException` to other place and just add a log here?

Reply:
Validation added